feat: add `stats/array/max-by`

gururaj1512 opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

-   feat: add `stats/array/max-by`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Needs #7129 to be merged

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add `stats/array/max-by`

gururaj1512 commented on lib/node_modules/@stdlib/stats/array/max-by/README.md:
```diff
@@ -0,0 +1,155 @@
+-   A provided callback function should return a numeric value.
+-   If a provided callback function does not return any value (or equivalently, explicitly returns `undefined`), the value is **ignored**.
+-   The function supports array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
+-   When possible, prefer using [`dmax`][@stdlib/stats/array/dmax], [`smax`][@stdlib/stats/array/smax], and/or [`max`][@stdlib/stats/array/max], as, depending on the environment, these interfaces are likely to be significantly more performant.
```
The packages `dmax` & `smax` are yet to be added.
Should they be mentioned here?
I think we can keep it as link attached would be same after addition of package.

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/README.md:
```diff
@@ -0,0 +1,155 @@
+-   A provided callback function should return a numeric value.
+-   If a provided callback function does not return any value (or equivalently, explicitly returns `undefined`), the value is **ignored**.
+-   The function supports array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
+-   When possible, prefer using [`dmax`][@stdlib/stats/array/dmax], [`smax`][@stdlib/stats/array/smax], and/or [`max`][@stdlib/stats/array/max], as, depending on the environment, these interfaces are likely to be significantly more performant.
```
```suggestion
```

We won't be adding those packages, so this comment should be removed in its entirety.

kgryte replied:
The `stats/array/*` namespace is intentionally only "generic".

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into stats-array-max-by

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/docs/repl.txt:
```diff
@@ -0,0 +1,44 @@
+
+{{alias}}( x )
```
```suggestion
{{alias}}( x, clbk[, thisArg] )
```

kgryte committed:
docs: fix signature

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/docs/repl.txt:
```diff
@@ -0,0 +1,44 @@
+
+    The callback function is provided three arguments:
+
+    - value: array element.
+    - aidx: array index.
+    - sidx: strided index (offset + aidx*stride).
+    - array: the input array.
```
This is a copy-paste mistake and should be updated.

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/README.md:
```diff
@@ -0,0 +1,155 @@
+-   **value**: array element.
+-   **aidx**: array index.
+-   **sidx**: strided index (`offset + aidx*stride`).
+-   **array**: input array/collection.
```
This isn't correct. But also, I see why you did this. I've updated the implementation to demonstrate how we address.

kgryte committed:
fix: update callback invocation logic to ensure correct arguments

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
style: fix indentation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
style: fix indentation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
I've updated the implementation. Docs and likely tests/examples should be updated accordingly.

gururaj1512 committed:
docs: update docs as suggested in docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/README.md:
```diff
@@ -0,0 +1,147 @@
+-   **clbk**: callback function.
+-   **thisArg**: execution context (_optional_).
+
+The invoked callback is provided four arguments:
+
+-   **value**: array element.
+-   **aidx**: array index.
```
```suggestion
The invoked callback is provided three arguments:

-   **value**: current array element.
-   **index**: current array index.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/README.md:
```diff
@@ -0,0 +1,147 @@
+
+-   **value**: current array element.
+-   **index**: current array index.
+-   **array**: input array/collection.
```
```suggestion
-   **array**: input array.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/docs/repl.txt:
```diff
@@ -0,0 +1,43 @@
+
+    The callback function is provided three arguments:
+
+    - value: array element.
+    - aidx: array index.
```
```suggestion
    - value: current array element.
    - index: current array index.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update parameters and descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: improve type specificity

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/lib/main.js:
```diff
@@ -0,0 +1,93 @@
+* @param {Callback} clbk - callback
+* @param {*} [thisArg] - execution context
+* @throws {TypeError} first argument must be an array-like object
+* @throws {TypeError} first argument must have a supported data type
```
```suggestion
* @throws {TypeError} first argument must have a supported data type
* @throws {TypeError} second argument must be a function
```

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/lib/main.js:
```diff
@@ -0,0 +1,93 @@
+		throw new TypeError( format( 'invalid argument. First argument must have one of the following data types: "%s". Data type: `%s`.', join( IDTYPES, '", "' ), dt ) );
+	}
+	if ( !isFunction( clbk ) ) {
+		throw new TypeError( format( 'invalid argument. Callback function must be a function. Value: `%s`.', clbk ) );
```
```suggestion
		throw new TypeError( format( 'invalid argument. Second argument must be a function. Value: `%s`.', clbk ) );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,274 @@
+
+	function badValue( value ) {
+		return function badValue() {
+			maxBy( value );
```
```suggestion
			maxBy( value, accessor );
```

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,274 @@
+
+	function badValue( value ) {
+		return function badValue() {
+			maxBy( value );
```
```suggestion
			maxBy( value, accessor );
```

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,274 @@
+	}
+});
+
+tape( 'the function throws an error if provided a callback function argument which is not a function', function test( t ) {
```
```suggestion
tape( 'the function throws an error if provided a second argument which is not a function', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,274 @@
+	}
+});
+
+tape( 'the function throws an error if provided a callback function argument which is not a function (options)', function test( t ) {
+	var values;
+	var i;
+	var x;
+
+	x = [ 1.0, -2.0, -4.0, 5.0, 0.0, 3.0 ];
+	values = [
+		'5',
+		5,
+		NaN,
+		null,
+		void 0,
+		true,
+		[],
+		{}
+	];
+
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			maxBy( x, value, {} );
+		};
+	}
+});
+
```
```suggestion
```

The function doesn't support any options.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,245 @@
+	t.end();
+});
+
+tape( 'if provided an array containing a single element, the function returns the first element applying callback function (accessors)', function test( t ) {
```
```suggestion
tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function', function test( t ) {
	var v = maxBy( [ 1.0 ], accessor );
	t.strictEqual( v, 2.0, 'returns expected value' );
	t.end();
});

tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function (accessors)', function test( t ) {
```

kgryte committed:
test: update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,245 @@
+	t.end();
+});
+
+tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function', function test( t ) {
+	var v = maxBy( [ 1.0 ], accessor );
+	t.strictEqual( v, 2.0, 'returns expected value' );
+	t.end();
+});
+
+tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function (accessors)', function test( t ) {
```
```suggestion
tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function to that element', function test( t ) {
	var v = maxBy( [ 1.0 ], accessor );
	t.strictEqual( v, 2.0, 'returns expected value' );
	t.end();
});

tape( 'if provided an array containing a single element, the function returns the result of applying a provided callback function to that element (accessors)', function test( t ) {
```

kgryte committed:
test: update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,245 @@
+		return v * 2.0;
+	}
+});
+
```
```suggestion

tape( 'the function supports providing a callback execution context (accessors)', function test( t ) {
	var ctx;
	var x;

	x = [ 1.0, 2.0, 3.0, 4.0, 5.0 ];
	ctx = {
		'count': 0
	};
	maxBy( toAccessorArray( x ), accessor, ctx );

	t.strictEqual( ctx.count, x.length, 'returns expected value' );
	t.end();

	function accessor( v ) {
		this.count += 1; // eslint-disable-line no-invalid-this
		return v * 2.0;
	}
});

```

kgryte committed:
test: add test

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: update callback tests

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (dismissed):
LGTM

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into stats-array-max-by

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,306 @@
+
+	expected = [ x, x, x, x, x ];
+	t.deepEqual( arrays, expected, 'returns expected value' );
+	
```
```suggestion

```

kgryte commented on lib/node_modules/@stdlib/stats/array/max-by/test/test.js:
```diff
@@ -0,0 +1,306 @@
+
+	expected = [ ax, ax, ax, ax, ax ];
+	t.deepEqual( arrays, expected, 'returns expected value' );
+	
```
```suggestion

```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>
